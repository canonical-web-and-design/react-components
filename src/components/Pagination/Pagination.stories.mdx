import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Pagination from "./Pagination";

<Meta title="Pagination" component={Pagination} />

### Pagination

This is a [React](https://reactjs.org/) component for the Vanilla [Pagination](https://docs.vanillaframework.io/patterns/pagination/).

The pagination component should be used to navigate between pages of content. Depending on the length provided, the pagination component will automatically scale.

### Props

<ArgsTable of={Pagination} />

### Default

<Canvas>
  <Story name="Default">
    <Pagination
      itemsPerPage={20}
      totalItems={100}
      paginate={() => {}}
      currentPage={3}
    />
  </Story>
</Canvas>

### Truncated

<Canvas>
  <Story name="Truncated">
    <Pagination
      itemsPerPage={10}
      totalItems={1000}
      paginate={() => {}}
      currentPage={34}
    />
  </Story>
</Canvas>

### Disabled controls

<Canvas>
  <Story name="Disabled controls">
    <Pagination
      itemsPerPage={20}
      totalItems={100}
      paginate={() => {}}
      currentPage={1}
    />
  </Story>
</Canvas>
