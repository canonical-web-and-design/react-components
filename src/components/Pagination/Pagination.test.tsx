import { shallow } from "enzyme";
import React from "react";

import Pagination from "./Pagination";

describe("<Pagination />", () => {
  // snapshot tests
  it("renders and matches the snapshot", () => {
    const component = shallow(
      <Pagination
        itemsPerPage={10}
        totalItems={50}
        paginate={jest.fn()}
        currentPage={1}
      />
    );

    expect(component).toMatchSnapshot();
  });

  // unit tests
  it("renders no pagination with only a single page", () => {
    const component = shallow(
      <Pagination
        itemsPerPage={10}
        totalItems={5}
        paginate={jest.fn()}
        currentPage={1}
      />
    );

    expect(component.find("nav").exists()).toBe(false);
  });

  it("renders a simple paginator with back and forward arrows if only five pages or less", () => {
    const component = shallow(
      <Pagination
        itemsPerPage={10}
        totalItems={50}
        paginate={jest.fn()}
        currentPage={1}
      />
    );

    expect(component.find("PaginationItemSeparator").exists()).toBe(false);
    expect(component.find("PaginationButton").length).toEqual(2);
    expect(component.find("PaginationItem").length).toEqual(5);
  });

  it("renders a complex paginator with truncation if more than five pages", () => {
    const component = shallow(
      <Pagination
        itemsPerPage={10}
        totalItems={1000}
        paginate={jest.fn()}
        currentPage={5}
      />
    );

    expect(component.find("PaginationItemSeparator").length).toEqual(2);
    expect(component.find("PaginationButton").length).toEqual(2);
    expect(component.find("PaginationItem").length).toEqual(5);
  });

  it("does not render a truncation separator if currentPage is contiguous at start", () => {
    const component = shallow(
      <Pagination
        itemsPerPage={10}
        totalItems={1000}
        paginate={jest.fn()}
        currentPage={2}
      />
    );

    expect(component.find("PaginationItemSeparator").length).toEqual(1);
    expect(component.find("PaginationButton").length).toEqual(2);
    expect(component.find("PaginationItem").length).toEqual(5);
  });

  it("does not render a truncation separator if currentPage is contiguous at end", () => {
    const component = shallow(
      <Pagination
        itemsPerPage={10}
        totalItems={1000}
        paginate={jest.fn()}
        currentPage={98}
      />
    );

    expect(component.find("PaginationItemSeparator").length).toEqual(1);
    expect(component.find("PaginationButton").length).toEqual(2);
    expect(component.find("PaginationItem").length).toEqual(5);
  });
});
