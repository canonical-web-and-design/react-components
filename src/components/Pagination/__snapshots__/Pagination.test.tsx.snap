// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders and matches the snapshot 1`] = `
<nav>
  <ul
    className="p-pagination"
  >
    <PaginationButton
      direction="back"
      disabled={true}
      key="back"
      onClick={[Function]}
    />
    <PaginationItem
      isActive={true}
      key="1"
      number={1}
      onClick={[Function]}
    />
    <PaginationItem
      isActive={false}
      key="2"
      number={2}
      onClick={[Function]}
    />
    <PaginationItem
      isActive={false}
      key="3"
      number={3}
      onClick={[Function]}
    />
    <PaginationItem
      isActive={false}
      key="4"
      number={4}
      onClick={[Function]}
    />
    <PaginationItem
      isActive={false}
      key="5"
      number={5}
      onClick={[Function]}
    />
    <PaginationButton
      direction="forward"
      disabled={false}
      key="forward"
      onClick={[Function]}
    />
  </ul>
</nav>
`;
